core/camel-support/src/main/java/org/apache/camel/support/EndpointHelper.java

```diff
@@ -126,24 +127,40 @@ public static boolean matchEndpoint(CamelContext context, String uri, String pat
         }
 
         // we need to test with and without scheme separators (//)
-        if (uri.contains("://")) {
-            // try without :// also
-            String scheme = StringHelper.before(uri, "://");
-            String path = after(uri, "://");
-            if (PatternHelper.matchPattern(scheme + ":" + path, pattern)) {
-                return true;
-            }
-        } else {
-            // try with :// also
-            String scheme = StringHelper.before(uri, ":");
-            String path = after(uri, ":");
-            if (PatternHelper.matchPattern(scheme + "://" + path, pattern)) {
-                return true;
+        boolean match = PatternHelper.matchPattern(toggleUriSchemeSeparators(uri), pattern);
+        match |= PatternHelper.matchPattern(uri, pattern);
+        if (!match && pattern != null && pattern.contains("?")) {
+            // try normalizing the pattern as a uri for exact matching, so parameters are ordered the same as in the endpoint uri
+            try {
+                pattern = URISupport.normalizeUri(pattern);
+                // try both with and without scheme separators (//)
+                match = toggleUriSchemeSeparators(uri).equalsIgnoreCase(pattern);
+                return match || uri.equalsIgnoreCase(pattern);
+            } catch (URISyntaxException e) {
```

Comment:
I dont like swallowed exceptions as they impact performance. We need to consider that matchEndpoint if its used during routing then this can slow down unnecessary.

Need to investigate some more

Reply:
Okay looked at this, and we only use this for interceptors at runtime so there is no normal overhead. And we also normalize the uri, so this logic does attempt to normalize also.